rust/driver/src/lib.rs

```diff
@@ -42,255 +46,319 @@ pub unsafe extern "C" fn get_tx_context_zeroed(_context: *mut ffi::c_void) -> ev
     TX_CONTEXT_ZEROED
 }
 
-/// # Safety
-///
-/// All pointers must be valid, except for `context` which can be null if the `evmc_host_interface`
-/// accepts null pointers as context.
-pub unsafe fn run_raw(
-    host: *const evmc_host_interface,
-    context: *mut ffi::c_void,
-    revision: Revision,
-    message: *const evmc_message,
-    code: *const u8,
-    code_len: usize,
-) -> evmc_result {
-    let instance = evmc_create_evmrs();
-    if instance.is_null() {
-        panic!("vm instance is null")
-    }
-    // SAFETY:
-    // `instance is not null`. `evmc_create_evmrs` must return a valid pointer to an `evmc_vm_t`.
-    let instance = unsafe { &mut *(instance as *mut evmc_vm_t) };
+pub struct Instance(&'static mut evmc_vm_t);
 
-    let execute = instance.execute.unwrap();
-    let destroy = instance.destroy.unwrap();
-
-    let result = unsafe { execute(instance, host, context, revision, message, code, code_len) };
-
-    unsafe {
-        destroy(instance);
+impl Deref for Instance {
+    type Target = evmc_vm_t;
+    fn deref(&self) -> &Self::Target {
+        self.0
     }
+}
 
-    result
+impl DerefMut for Instance {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
 }
 
-pub fn run<T>(
-    host: &evmc_host_interface,
-    context: &mut T,
-    revision: Revision,
-    message: &evmc_message,
-    code: &[u8],
-) -> evmc_result {
-    // SAFETY:
-    // All pointer are valid since they are created from references.
-    unsafe {
-        run_raw(
-            host,
-            context as *mut T as *mut ffi::c_void,
-            revision,
-            message,
-            if code.is_empty() {
-                ptr::null()
-            } else {
-                code.as_ptr()
-            },
-            code.len(),
-        )
+impl Default for Instance {
+    fn default() -> Self {
+        let instance = unsafe { evmc_create_evmrs() };
+        if instance.is_null() {
+            panic!("vm instance is null")
```

Comment:
Messages read better when they are in the form:
`failed to do X` (+  optionally `, reason is Y`)
here message would be better: `failed to construct default VM instance`